src/System.Linq.Expressions/src/System/Linq/Expressions/Compiler/LambdaCompiler.Generated.cs

```diff
@@ -12,6 +14,24 @@ partial class LambdaCompiler
         [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Maintainability", "CA1505:AvoidUnmaintainableCode"), System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Maintainability", "CA1502:AvoidExcessiveComplexity")]
         private void EmitExpression(Expression node, CompilationFlags flags)
         {
+            // When compling deep trees, we run the risk of triggering a terminating StackOverflowException,
+            // so we probe for sufficient stack here. When we run out of stack, we simply continue the work
+            // on another thread. Note that EnsureSufficientExecutionStack is very fast, so the overhead of
+            // this overflow protection is minimal.
+            try
+            {
+                RuntimeHelpers.EnsureSufficientExecutionStack();
+            }
+            catch (InsufficientExecutionStackException)
+            {
+                Task.Factory.StartNew(s =>
+                {
+                    var t = (Tuple<Expression, CompilationFlags>)s;
+                    EmitExpression(t.Item1, t.Item2);
+                }, Tuple.Create(node, flags)).Wait();
```

Comment:
A few things here:
1. When Wait'ing on a Task, if the task hasn't yet been executed elsewhere, the Wait will ask the scheduler if it's ok to just run the work on the current thread.  If it says yes, then this Wait will end up inlining the work onto this thread, defeating the purpose of this offloading.
2. For better or worse, Task.Factory.StartNew targets the current scheduler, whereas you probably want to target the default scheduler.
3. Wait() will wrap any exceptions that occur in an AggregateException, but you probably just want the original exception propagated.
4. EmitExpression is an instance method, so even though you're trying to avoid a closure, you're still getting one.
Instead I think you want something more like:
```C#
try
{
    RuntimeHelpers.EnsureSufficientExecutionStack();
}
catch (InsufficientExecutionStackException)
{
    Task task = Task.Factory.StartNew(s =>
    {
        var t = (Tuple<LambdaCompiler, Expression, CompilationFlags>)s;
        t.Item1.EmitExpression(t.Item2, t.Item3); // addresses #4
}, Tuple.Create(this, node, flags), CancellationToken.None, TaskCreationOptions.DenyChildAttach, TaskScheduler.Default); // addresses #2
((IAsyncResult)task).AsyncWaitHandle.WaitOne(); // addresses #1
task.GetAwaiter().GetResult(); // addresses #3
return;
```
If it won't impact perf to much, it might also be nice to encapsulate this into a helper that you can use instead of copying this pattern in all places below.